src/nvim/lua/executor.c

```diff
@@ -1008,3 +1012,253 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+static int nlua_http_parser_execute(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  http_parser *p = luaL_checkudata(lstate, 1, "http_parser");
+  size_t chunk_len = 0;
+  const char *chunk = luaL_checklstring(lstate, 2, &chunk_len);
+  lua_getfenv(lstate, 1);
+  size_t bytes_parsed
+      = http_parser_execute(p, &lua_http_parser_settings, chunk, chunk_len);
+  lua_pop(lstate, 1);
+  if (p->http_errno) {
+    // TODO(ashkan): use the http_errno_name() somehow?
+    lua_pushstring(lstate, http_errno_description(p->http_errno));
+    return lua_error(lstate);
+  }
+
+  lua_Integer ret = (lua_Integer)bytes_parsed;
+  lua_pushinteger(lstate, ret);
+
+  // TODO(ashkan): do something if this is finished?
+  // if (http_body_is_final(p)) {
+  //   lua_getfenv(lstate, 1);
+  //   return 2;
+  // }
+
+  return 1;
+}
+
+// TODO(ashkan): improve diagnostics?
+static int nlua_http_parser_to_string(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  lua_pushliteral(lstate, "<http_parser>");
+  return 1;
+}
+
+static int nlua_http_parser_index(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+{
+  luaL_checkudata(lstate, 1, "http_parser");
+  lua_getfenv(lstate, 1);
+  // Return the whole table if the key is "table"
+  if (0 == strcmp(luaL_checkstring(lstate, 2), "table")) {
+    return 1;
+  }
+  lua_pushvalue(lstate, 2);
+  lua_rawget(lstate, -2);
+  lua_remove(lstate, -2);
+  return 1;
+}
+
+static int nlua_create_http_parser(lua_State *const lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  http_parser *p = lua_newuserdata(lstate, sizeof(http_parser));  // [result]
+  http_parser_init(p, HTTP_BOTH);
+  luaL_newmetatable(lstate, "http_parser");        // [result, meta]
+  lua_setmetatable(lstate, -2);                    // [result]
+  lua_newtable(lstate);                            // [result, fenv]
+  lua_newtable(lstate);                            // [result, fenv, headers]
+  lua_setfield(lstate, -2, LUA_HTTP_HEADERS_KEY);  // [result, fenv]
+  lua_setfenv(lstate, -2);                         // [result]
+  return 1;
+}
+
+static int nlua_http_status_name(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+{
+  long status_code = luaL_checkinteger(lstate, 1);
+  lua_pushstring(lstate, http_status_str((enum http_status)status_code));
+  return 1;
+}
+
+static const char *HTTP_FIELD_NAMES[UF_MAX] = {
+    "schema", "host", "port", "path", "query", "fragment", "userinfo",
+};
+
+// nlua_http_parse_url(url, [is_connect])
+static int nlua_http_parse_url(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
+{
+  size_t input_len = 0;
+  const char *input = luaL_checklstring(lstate, 1, &input_len);
+  int is_connect = (int)lua_toboolean(lstate, 2);
+
+  struct http_parser_url u;
+  http_parser_url_init(&u);
+  if (http_parser_parse_url(input, input_len, is_connect, &u) != 0) {
+    // We would give a more informative error, but http_parser doesn't give us
+    // any.
+    return luaL_error(lstate, "Failed to parse url.");
```

Comment:
we generally avoid punctuation in error messages where possible. So a period is inconsistent with most (should be _all_) other messages.

```suggestion
    return luaL_error(lstate, "Failed to parse url");
```